storage/tests/curl_request_integration_test.cc

```diff
@@ -108,10 +114,55 @@ TEST(CurlRequestTest, SimplePOST) {
   request.AddHeader("charsets: utf-8");
 
   request.PrepareRequest(data);
-  auto response_text = request.MakeRequest();
-  nl::json response = nl::json::parse(response_text);
-  nl::json form = response["form"];
+  auto response = request.MakeRequest();
+  EXPECT_EQ(200, response.status_code);
+  nl::json parsed = nl::json::parse(response.payload);
+  nl::json form = parsed["form"];
   EXPECT_EQ("foo1&foo2 foo3", form["foo"].get<std::string>());
   EXPECT_EQ("bar1-bar2", form["bar"].get<std::string>());
   EXPECT_EQ("baz=baz2", form["baz"].get<std::string>());
 }
+
+TEST(CurlRequestTest, Handle404) {
+  storage::internal::CurlRequest request(
+      "https://nghttp2.org/httpbin/status/404");
```

Comment:
Hmm, this is actually making external network calls? This is not a unit test anymore. :-(

Can we mock the reply and parse it locally without having to use the network here?

Reply:
And that is why it is called an integration test?

Reply:
Even if it's an integration test, it depends on a third-party service running elsewhere: it could be down, or have an error, or have high latency, and cause these tests to fail. You know that this would not be cool in google3. :-)

If you'd like a real integration test, you can run your own local webserver with 0 code via Python, e.g., see https://docs.python.org/2/library/simplehttpserver.html (see the command line to start a simple web server).

It will get you the 200 and the 404 results easily, not sure about the custom headers or teapot, you may have to add that yourself, but it would at least be more hermetic.

Reply:
Okay, I hear you.  I created #542 to track this.

Reply:
Please add TODOs for each of the remote URL fetches with the issue number so that it's easy to find them?